### ISSUE TITLE
serialization of function groups

### ISSUE THREAD

**Issue by apex8** [larshp/abapGit#914]
When pulling my project into another system I noticed missing include files from a function group. The missing include files are not even in the source repository.

I have a generated Function Group for my generated maintenance views. On the source system I have the files

- /namespace/lcrt_genf00
- /namespace/lcrt_genf01 (which isn´t used in the FG main program
- /namespace/lcrt_geni00
- /namespace/lcrt_gent00
- /namespace/lcrt_gentop
- /namespace/lcrt_genuxx
- [tons of other FM-includes that are included in uxx]

After cloning the repository in the other system the includes F01 (unused) and T00 (included in TOP) are missing.

**Comment by larshp** [comment]
Hi,
Can you try with a functiongroup without namespace, does it work?

**Comment by apex8** [comment]
Hi,
I created some useless FGroup and tested it with a local repo. I looked a bit deeper in this and found the FM RS_GET_ALL_INCLUDES, which is used to determine all the includes of a FGroup at the beginning of the serialization.
The interesting thing is that this T00 Include is not provided by that FM. I also played a bit with the FM parameter, but that didn´t help.

**Comment by apex8** [comment]
There is another thing I noticed. There´s always a UXX include in a FGroup where all the include commands for the FMs are included. That UXX include isn´t part of the include list as well, but this doesn´t seem to be an issue when pulling the project.

**Comment by apex8** [comment]
I tried the same with a FGroup without any namespace (just Z), added a database table to create some maintenance view. I had some T00 include in the FGroup afterwards. Then I executed FM RS_GET_ALL_INCLUDES for that FGroup - T00 is missing. Looked for some notes and tried with a newer system (SAP_BASIS 740 SP14) - same problem

**Comment by apex8** [comment]
SAP development said the FM works as intended. The reason they gave was that L<FGroupname>T## includes are generated.

**Comment by larshp** [comment]
Hi, 
Just installed https://github.com/Yelcho/sapui5-deployer without any problems

Are there any problems except the missing includes? Does the installed function group work, if it works then I suggest we close this issue?

**Comment by apex8** [comment]
Yes there are problems. It cannot be activated. It was an Fgroup for
generated code of maintenance views, so those could just be regenerated,
but I wouldn`t consider it as a good solution for those fgroups. Since SAP
said that the fm for collecting the includes works as intended, maybe that
task should be done manually to cover the generated includes.

However there was another problem: When activating I got errors stating
that the relation of some fm`s to the fgroups is ambiguous.  In the long
text some cut part of the fgroups name was mentioned as the other fgroup.
My assumption on this would be that this could be some issue with the
namespace I use.
I can provide more details on Monday.

Regards
Alexander

On 9 Sep 2017 4:09 PM, "Lars Hvam" <notifications@github.com> wrote:

Hi,
Just installed https://github.com/Yelcho/sapui5-deployer without any
problems

Are there any problems except the missing includes? Does the installed
function group work, if it works then I suggest we close this issue?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/larshp/abapGit/issues/914#issuecomment-328279569>, or mute
the thread

**Comment by larshp** [comment]
Hi,
Can you help making an example in a public repository? Where the activation fails?

Note that abapGit is a community project, so do not expect any support from SAP, most of the functionality in abapGit uses unsupported/unreleased APIs and functions

**Comment by apex8** [comment]
I'll prepare an public repository. Regarding support, I have the
possibility to create support tickets and I use it. Usually they do not ask
for what I use their stuff, and if they answer that's good enough for me

**Comment by apex8** [comment]
okay there it is:
I created two repositories:
[FGroupTest](https://github.com/apex8/fgrouptest) - which is a 'Z- Function Group' in a local package. This one could be deserialized without a problem; the T00 include is part of the serialization.

[FGroupTestNS](https://github.com/apex8/fgrouptest_ns) - which is a 'Namespace - Function Group' in a transport related package.
Deserialization terminates with error, because the T00 include is missing; its also not in the repository/wasn´t serialized

**Comment by larshp** [comment]
Thanks, I do not have development access keys keys to test namespaces, but pull requests are welcome.

**Comment by apex8** [comment]
Yeah I figured that. 
The interesting part here should be the serialization. I´ll have a look.

**Comment by apex8** [comment]
created PR [#924](https://github.com/larshp/abapGit/pull/924)

**Comment by apex8** [comment]
OK files seem to be ok now. 
FGroups involving namespaces now have an activation error. 
The code looks totally okay, but the error message says that the code of the FM (U01 include) is not accessible. However the U01 include is placed in the UXX include which is placed in the FGroup Main program.

I tried regenerating the maintenance screen which results in the error: "Function group /<namespace>/DBFGTST ambiguous".
There is some extensive longtext, which starts with:
The function group **/BITGRP/DBFGTST** is passed by the control catalog to the generation. The function group **/BITGRP/DBFGTST** is not identical to the function group **/BIT**, to which the function modules are actually assigned, according to the function library.

It seems to be like some sort of registration issue, as there is another function group mentioned which is an incomplete part of the correct one. Currently I don´t know where to look for that, but as soon I find some more time, I try to investigate.

**Comment by apex8** [comment]
submitted PR #925

**Comment by apex8** [comment]
Today I was able to pull my whole project successfully on a second server. However as I had issues with deserializing Function Groups in the past, its still the best to delete a "partly" existing Function Group. If I do not delete the Group before pulling, I repeatedly run into some error stating that my function Group has more than 99 includes/function modules. So as the deserialization is deleting every function module include before creating it new, this seems to fail in some cases. I found multiple FM-includes for the same FM.
That´s why I would like to keep this issue open and do some more testing.
